cmd/thanos/remote-write-receive.go

```diff
@@ -0,0 +1,281 @@
+	dataDir string,
+	peer cluster.Peer,
+	component string,
+) error {
+	level.Info(logger).Log("msg", "setting up receiver")
+
+	var metadata = &promMetadata{
+		// Start out with the full time range. The shipper will constrain it later.
+		// TODO(fabxc): minimum timestamp is never adjusted if shipping is disabled.
+		mint: 0,
+		maxt: math.MaxInt64,
+	}
+
+	tsdbCfg := &tsdb.Options{
+		Retention:        model.Duration(time.Hour * 24 * 15),
+		NoLockfile:       true,
+		MinBlockDuration: model.Duration(time.Hour * 2),
+		MaxBlockDuration: model.Duration(time.Hour * 2),
+	}
+
+	ctxWeb, cancelWeb := context.WithCancel(context.Background())
+	localStorage := &tsdb.ReadyStorage{}
+	receiver := receive.NewReceiveWriter(log.With(logger, "component", "receive-writer"), localStorage)
+	webHandler := receive.NewHandler(log.With(logger, "component", "remote-receive-handler"), &receive.Options{
+		Context:        ctxWeb,
+		Receiver:       receiver,
+		ListenAddress:  remoteWriteAddress,
+		MaxConnections: 1000,
+		ReadTimeout:    time.Minute * 5,
+		Registry:       reg,
+		ReadyStorage:   localStorage,
+	})
+
+	// Start all components while we wait for TSDB to open but only load
+	// initial config and mark ourselves as ready after it completed.
+	dbOpen := make(chan struct{})
+
+	// sync.Once is used to make sure we can close the channel at different execution stages(SIGTERM or when the config is loaded).
+	type closeOnce struct {
+		C     chan struct{}
+		once  sync.Once
+		Close func()
+	}
+	// Wait until the server is ready to handle reloading.
+	reloadReady := &closeOnce{
+		C: make(chan struct{}),
+	}
+	reloadReady.Close = func() {
+		reloadReady.once.Do(func() {
+			close(reloadReady.C)
+		})
+	}
+
+	level.Debug(logger).Log("msg", "setting up joining cluster")
+	{
+		ctx, cancel := context.WithCancel(context.Background())
+		g.Add(func() error {
+			mint, maxt := metadata.Timestamps()
+			if peer != nil {
+				if err := peer.Join(cluster.PeerTypeSource, cluster.PeerMetadata{
+					Labels:  metadata.LabelsPB(),
+					MinTime: mint,
+					MaxTime: maxt,
+				}); err != nil {
+					return errors.Wrap(err, "join cluster")
+				}
+			}
+
+			<-ctx.Done()
+			return nil
+		}, func(err error) {
+			level.Debug(logger).Log("msg", "mesh group errored", "err", err)
+			cancel()
+			peer.Close(2 * time.Second)
+		})
+	}
+
+	level.Debug(logger).Log("msg", "setting up endpoint readiness")
+	{
+		// Initial configuration loading.
+		cancel := make(chan struct{})
+		g.Add(
+			func() error {
+				select {
+				case <-dbOpen:
+					break
+				// In case a shutdown is initiated before the dbOpen is released
+				case <-cancel:
+					reloadReady.Close()
+					return nil
+				}
+
+				reloadReady.Close()
+
+				webHandler.Ready()
+				level.Info(logger).Log("msg", "Server is ready to receive web requests.")
+				<-cancel
+				return nil
+			},
+			func(err error) {
+				level.Debug(logger).Log("msg", "initial load group errored", "err", err)
```

Comment:
hmm, I don't get this log messages here and on every group. This code is invoked even though this group did not errof (but e.g other group errored). I think we can remove those debug logs. We will print error for problematic group via `g.Run` anyway.

Reply:
Sorry, this is a leftover from some debugging, I'll remove it.